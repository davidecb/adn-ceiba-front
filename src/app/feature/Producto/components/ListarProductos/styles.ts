import styled from 'styled-components';

export const ProductosContainer = styled.div`
  width: 100%;
  display: flex;
  flex-wrap: wrap;
  justify-content: center;
`;

export const FormularioContainer = styled.div`
  position: fixed;
  top: 0;
  z-index: 100;
  background: rgba(0, 0, 0, 0.95);
  width: 100%;
  height: 100%;
  display: flex;
  justify-content: center;
  align-items: center;
`;
