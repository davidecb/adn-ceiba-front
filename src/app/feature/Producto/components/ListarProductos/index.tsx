import * as PropTypes from 'prop-types';
import * as React from 'react';
import { FormularioContainer, ProductosContainer } from './styles';
import { FormSolicitarProducto } from 'app/shared/components/FormSolicitarProducto';
import { Producto } from '../../models/Producto';
import { ProductoSolicitado } from '../../models/ProductoSolicitado';
import { TarjetaProducto } from '../TarjetaProducto';

export interface ListarProductosProps {
  productos: Array<Producto>;
  onClickAgregarProductoSolicitado: (productoSolicitado: ProductoSolicitado) => void;
}

export const ListarProductos: React.FC<ListarProductosProps> = ({
  productos,
  onClickAgregarProductoSolicitado,
}) => {
  const prod: Producto = {
    'id': 0,
    'nombre': '',
    'costo': 0,
    'tiempo': 0,
    'imagen': ''
};
  const [personalizarClicked, setPersonalizarClicked] = React.useState(false);
  const [productoSeleccionado, setProductoSeleccionado] = React.useState(prod);
  return (
    <ProductosContainer>
      {productos && productos.map((producto, index) => {
        return (
          <TarjetaProducto 
              key={index} 
              producto={producto}
              onClickPersonalizarProducto={()=>{
                setPersonalizarClicked(true);
                setProductoSeleccionado(producto);
              }}/>);
      })}
      {personalizarClicked && 
        <FormularioContainer>
          <FormSolicitarProducto 
            producto={productoSeleccionado} 
            onSubmit={(productoSolicitado) => {
              onClickAgregarProductoSolicitado(productoSolicitado); 
            }}
            cerrarFormulario={() => {
              setPersonalizarClicked(false);
              setProductoSeleccionado(prod);
            }}
          />
        </FormularioContainer>      
      }
    </ProductosContainer>
  );
};

ListarProductos.propTypes = {
  productos: PropTypes.array.isRequired,
  onClickAgregarProductoSolicitado: PropTypes.func.isRequired,
};
