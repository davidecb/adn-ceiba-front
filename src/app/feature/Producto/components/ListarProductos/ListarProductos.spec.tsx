import * as React from 'react';
import { RenderResult, fireEvent, render, wait } from '@testing-library/react';
import { SinonStub, stub } from 'sinon';
import { ListarProductos } from './../ListarProductos';

describe('ListarProductos Test', () => {
  let componentWrapper: RenderResult;
  let componentProps: React.ComponentProps<typeof ListarProductos> & {
    onClickAgregarProductoSolicitado: SinonStub;
  };

  beforeEach(() => {
    componentProps = {
      productos: [{
        id: 1,
        nombre: 'producto testing',
        costo: 10000,
        tiempo: 30,
        imagen: 'imagen.jpg'
      },{
        id: 2,
        nombre: 'producto testing 2',
        costo: 20000,
        tiempo: 60,
        imagen: 'imagen2.jpg'
      }],
      onClickAgregarProductoSolicitado: stub(),
    };
    componentWrapper = render(<ListarProductos {...componentProps} />);
  });

  it('should match snapshot', () => {
    expect(componentWrapper.container).toMatchSnapshot();
  });

  it('should show formSolicitarProducto on personalizarProducto click', async () => {
    const elem = componentWrapper.container;
    const personalizarButton = elem.querySelectorAll('button')[0];
    
    await wait(() => {
      personalizarButton && fireEvent.click(personalizarButton);
    });

    const divs = elem.querySelectorAll('div');
    const forms = elem.querySelectorAll('form');
    const h2s = elem.querySelectorAll('h2');

    expect(divs.length).toBe(16);
    expect(forms.length).toBe(1);
    expect(h2s.length).toBe(1);
    expect(h2s[0].textContent).toBe('Personalizar producto testing');
  });

  it('should close formSolicitarProducto on cancelar button click', async () => {
    const elem = componentWrapper.container;
    const personalizarButton = elem.querySelectorAll('button')[0];
    
    await wait(() => {
      personalizarButton && fireEvent.click(personalizarButton);
    });

    const cancelarButton = elem.querySelectorAll('span')[2];
    expect(cancelarButton.textContent).toBe('❌');

    await wait(() => {
      cancelarButton && fireEvent.click(cancelarButton);
    });

    const divs = elem.querySelectorAll('div');
    const forms = elem.querySelectorAll('form');

    expect(divs.length).toBe(9);
    expect(forms.length).toBe(0);
  });
  
  it('should submit and close the form when click on agregar al carrito button', async () => {
    const elem = componentWrapper.container;
    const personalizarButton = elem.querySelectorAll('button')[0];
    
    await wait(() => {
      personalizarButton && fireEvent.click(personalizarButton);
    });
    
    const material = elem.querySelector('select[name="material"]');
    const color = elem.querySelector('select[name="color"]');
    const submitButton = elem.querySelector('button[type="submit"]');

    await wait(() => {
      material && fireEvent.change(material, { target: { value: 'ABS'}});
    });
    await wait(() => {
      color && fireEvent.change(color, { target: { value: 'madera'}});
    });

    await wait(() => {
      submitButton && fireEvent.click(submitButton);
    });

    const divs = elem.querySelectorAll('div');
    const forms = elem.querySelectorAll('form');

    expect(divs.length).toBe(9);
    expect(forms.length).toBe(0);
  });
});

export {};
