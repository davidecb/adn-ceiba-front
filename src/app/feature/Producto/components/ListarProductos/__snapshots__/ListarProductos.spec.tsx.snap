// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListarProductos Test should match snapshot 1`] = `
<div>
  <div
    class="sc-AxjAm dPoJzv"
  >
    <div
      class="sc-fznyAO cEMFXg"
    >
      <img
        class="sc-fzqBZW bbHHRl"
        src="imagen.jpg"
      />
      <div
        class="sc-fznKkj FgPbq"
      >
        producto testing
      </div>
      <div
        class="sc-fznZeY gQhwXl"
      >
        $ 
        10000
      </div>
      <div
        class="sc-fzokOt jldLkq"
      >
        30
         min.
      </div>
      <button
        class="sc-AxmLO hpLBoc"
      >
        <span>
          personalizar
        </span>
      </button>
    </div>
    <div
      class="sc-fznyAO cEMFXg"
    >
      <img
        class="sc-fzqBZW bbHHRl"
        src="imagen2.jpg"
      />
      <div
        class="sc-fznKkj FgPbq"
      >
        producto testing 2
      </div>
      <div
        class="sc-fznZeY gQhwXl"
      >
        $ 
        20000
      </div>
      <div
        class="sc-fzokOt jldLkq"
      >
        60
         min.
      </div>
      <button
        class="sc-AxmLO hpLBoc"
      >
        <span>
          personalizar
        </span>
      </button>
    </div>
  </div>
</div>
`;
